libeos-updater-util/flatpak.c

```diff
@@ -1457,6 +1457,702 @@ euu_filter_for_existing_flatpak_ref_actions (GHashTable *ref_actions  /* (elemen
+{
+  gboolean related_ref_is_installed = FALSE;
+
+  g_return_val_if_fail (out_action_type != NULL, FALSE);
+
+  related_ref_is_installed = g_hash_table_contains (installed_related_refs, related_ref);
+
+  switch (source_ref_action->type)
+    {
+      case EUU_FLATPAK_REMOTE_REF_ACTION_INSTALL:
+        if (flatpak_related_ref_should_download (related_ref))
+          {
+            if (related_ref_is_installed)
+              *out_action_type = EUU_FLATPAK_REMOTE_REF_ACTION_UPDATE;
+            else
+              *out_action_type = EUU_FLATPAK_REMOTE_REF_ACTION_INSTALL;
+
+            return TRUE;
+          }
+        break;
+      case EUU_FLATPAK_REMOTE_REF_ACTION_UPDATE:
+        if (flatpak_related_ref_should_download (related_ref) &&
+            related_ref_is_installed)
+          {
+            *out_action_type = EUU_FLATPAK_REMOTE_REF_ACTION_UPDATE;
+            return TRUE;
+          }
+        break;
+      case EUU_FLATPAK_REMOTE_REF_ACTION_UNINSTALL:
+        if (flatpak_related_ref_should_delete (related_ref) &&
+            related_ref_is_installed)
+          {
+            *out_action_type = EUU_FLATPAK_REMOTE_REF_ACTION_UNINSTALL;
+            return TRUE;
+          }
+          break;
+      default:
+        break;
+    }
+
+  /* Nothing to do */
+  return FALSE;
+}
+
+static gboolean
+recursively_find_remote_and_installed_related_refs (FlatpakInstallation  *installation,
+                                                    FlatpakRef           *source_ref,
+                                                    GPtrArray            *remotes,
+                                                    GPtrArray            *remote_related_refs,
+                                                    GHashTable           *installed_related_refs,
+                                                    GCancellable         *cancellable,
+                                                    GError              **error)
+{
+  if (!populate_installed_related_refs (installation,
+                                        source_ref,
+                                        remotes,
+                                        installed_related_refs,
+                                        cancellable,
+                                        error))
+    return FALSE;
+
+  if (!populate_related_refs_in_all_remotes (installation,
+                                             source_ref,
+                                             remote_related_refs,
+                                             cancellable,
+                                             error))
+    return FALSE;
+
+  return TRUE;
+}
+
+static void
+initially_populate_remote_and_installed_related_refs (GPtrArray  **out_remote_related_refs,
+                                                      GHashTable **out_installed_related_refs,
+                                                      GPtrArray   *remotes)
```

Comment:
Nitpick: Typically, the ‘out’ arguments are listed after the ‘in’ ones. So I think `remotes` should be first here.